lib/src/test/java/org/kpipe/consumer/OffsetManagerTest.java

```diff
@@ -0,0 +1,798 @@
+        try {
+          final var cmd = commandQueue.poll(30, TimeUnit.SECONDS);
+          if (cmd != null) {
+            commandRef.set(cmd);
+            commandLatch.countDown();
+          }
+        } catch (final InterruptedException e) {
+          // Ignore interruption
+        }
+      });
+
+    // Start the commit operation
+    CompletableFuture.runAsync(() -> {
+      try {
+        offsetManager.commitSyncAndWait(30);
+      } catch (final InterruptedException e) {
+        // Ignore interruption
+      }
+    });
+
+    // Wait for command
+    final var commandReceived = commandLatch.await(10, TimeUnit.SECONDS);
+    assertTrue(commandReceived, "Command should be produced within timeout");
+
+    final var command = commandRef.get();
+    assertNotNull(command, "Commit command should be generated");
+
+    return command;
+  }
+
+  private void assertEventually(final Supplier<Boolean> condition, final Duration timeout, final String message)
+    throws Exception {
+    final var deadline = System.currentTimeMillis() + timeout.toMillis();
+
+    while (System.currentTimeMillis() < deadline) {
+      if (condition.get()) {
+        return;
+      }
+      Thread.sleep(50);
+    }
+
+    fail(message);
+  }
+
+  static Stream<Arguments> contiguousOffsetScenarios() {
+    return Stream.of(
+      Arguments.of(
+        "Basic gap scenario - process offsets out of order with a missing offset",
+        100L, // initialOffset
+        List.of(101L, 105L, 102L, 103L, 104L), // offsetsToTrack
+        List.of(102L, 101L, 104L), // offsetsToProcess (note: 103 is deliberately not processed)
+        103L // expectedCommittedOffset
+      ),
+      Arguments.of(
+        "Multiple gaps scenario - commit should stop at first unprocessed offset",
+        200L, // initialOffset
+        List.of(201L, 202L, 205L, 206L, 203L, 204L), // offsetsToTrack
+        List.of(201L, 202L, 203L), // offsetsToProcess (note: 204-206 are not processed)
+        204L // expectedCommittedOffset - commits up to the first gap
+      ),
+      Arguments.of(
+        "Large range with interleaved processing",
+        1000L, // initialOffset
+        List.of(1_001L, 1_002L, 1_010L, 1_020L, 1_030L, 1_005L, 1_006L), // offsetsToTrack
+        List.of(1_001L, 1_006L, 1_002L, 1_030L, 1_010L), // offsetsToProcess (note: 1005L and 1020L not processed)
+        1005L // expectedCommittedOffset - commits up to 1005L (the first gap is at 1005L)
+      )
+    );
+  }
+
+  static Stream<Arguments> gapSizeScenarios() {
+    return Stream.of(
+      Arguments.of(
+        "Large gap between offsets",
+        1_000L, // initial offset
+        5_000L, // gap offset
+        1_001L, // expected offset after first commit
+        5_001L // expected offset after the second commit
+      ),
+      Arguments.of(
+        "Backwards gap (higher offset processed before lower)",
+        5_000L, // initial offset
+        1_000L, // gap offset (lower than initial - backwards)
```

Comment:
this should be impossible - Kafka ordering guarantees (and the fact that track offset is called from main / single thread) make it invalid...

Reply:
removed